File: backend/app/api/tasks.py
from datetime import date
from typing import List, Optional
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.user import User
from app.models.task import TaskStatus
from app.schemas.task import (
    TaskCreate, TaskUpdate, TaskResponse, TaskSummaryResponse,
    SnoozeRequest, RescheduleRequest, NotificationIdUpdate
)
from app.services.task_service import TaskService

router = APIRouter(prefix="/tasks", tags=["Tasks"])


@router.get("/summary", response_model=TaskSummaryResponse)
async def get_task_summary(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    summary = await TaskService.get_task_summary(db, current_user.id)
    return TaskSummaryResponse(**summary)


@router.get("", response_model=List[TaskResponse])
async def list_tasks(
    status: Optional[TaskStatus] = Query(None, description="Filter by status"),
    search: Optional[str] = Query(None, description="Search in title or description"),
    date_filter: Optional[date] = Query(None, description="Filter by task start date"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    tasks = await TaskService.list_tasks(
        db=db,
        user_id=current_user.id,
        status_filter=status,
        search=search,
        date_filter=date_filter
    )
    return [TaskResponse.model_validate(t) for t in tasks]


@router.post("", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
async def create_task(
    task_in: TaskCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.create_task(db, current_user.id, task_in)
    return TaskResponse.model_validate(task)


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.get_task_by_id(db, task_id, current_user.id)
    return TaskResponse.model_validate(task)


@router.put("/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: int,
    task_in: TaskUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.update_task(db, task_id, current_user.id, task_in)
    return TaskResponse.model_validate(task)


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    await TaskService.delete_task(db, task_id, current_user.id)
    return None


@router.post("/{task_id}/pause", response_model=TaskResponse)
async def pause_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.pause_task(db, task_id, current_user.id)
    return TaskResponse.model_validate(task)


@router.post("/{task_id}/resume", response_model=TaskResponse)
async def resume_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.resume_task(db, task_id, current_user.id)
    return TaskResponse.model_validate(task)


@router.post("/{task_id}/snooze", response_model=TaskResponse)
async def snooze_task(
    task_id: int,
    snooze_in: SnoozeRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.snooze_task(db, task_id, current_user.id, snooze_in)
    return TaskResponse.model_validate(task)


@router.post("/{task_id}/reschedule", response_model=TaskResponse)
async def reschedule_task(
    task_id: int,
    reschedule_in: RescheduleRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.reschedule_task(db, task_id, current_user.id, reschedule_in)
    return TaskResponse.model_validate(task)


@router.post("/{task_id}/complete", response_model=TaskResponse)
async def complete_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.complete_task(db, task_id, current_user.id)
    return TaskResponse.model_validate(task)


@router.patch("/{task_id}/notification", response_model=TaskResponse)
async def update_notification_id(
    task_id: int,
    payload: NotificationIdUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    task = await TaskService.get_task_by_id(db, task_id, current_user.id)
    task.notification_id = payload.notification_id
    await db.commit()
    await db.refresh(task)
    return TaskResponse.model_validate(task)

File: frontend/src/api/tasks.ts
import { apiClient } from './client';
import { 
  Task, 
  TaskCreatePayload, 
  TaskUpdatePayload, 
  TaskSummary, 
  TaskStatus,
  SnoozePayload,
  ReschedulePayload,
  PlanDayRequest,
  PlanDayResponse
} from '../types';

export const tasksApi = {
  async planDay(payload: PlanDayRequest): Promise<PlanDayResponse> {
    const res = await apiClient.post<PlanDayResponse>('/tasks/plan-day', payload);
    return res.data;
  },

  async createBatchTasks(tasks: TaskCreatePayload[]): Promise<Task[]> {
    const res = await apiClient.post<Task[]>('/tasks/batch', { tasks });
    return res.data;
  },
  async getTasks(params?: { status?: TaskStatus; search?: string; date_filter?: string }): Promise<Task[]> {
    const res = await apiClient.get<Task[]>('/tasks', { params });
    return res.data;
  },

  async getSummary(): Promise<TaskSummary> {
    const res = await apiClient.get<TaskSummary>('/tasks/summary');
    return res.data;
  },

  async getTask(id: number): Promise<Task> {
    const res = await apiClient.get<Task>(`/tasks/${id}`);
    return res.data;
  },

  async createTask(payload: TaskCreatePayload): Promise<Task> {
    const res = await apiClient.post<Task>('/tasks', payload);
    return res.data;
  },

  async updateTask(id: number, payload: TaskUpdatePayload): Promise<Task> {
    const res = await apiClient.put<Task>(`/tasks/${id}`, payload);
    return res.data;
  },

  async deleteTask(id: number): Promise<void> {
    await apiClient.delete(`/tasks/${id}`);
  },

  async pauseTask(id: number): Promise<Task> {
    const res = await apiClient.post<Task>(`/tasks/${id}/pause`);
    return res.data;
  },

  async resumeTask(id: number): Promise<Task> {
    const res = await apiClient.post<Task>(`/tasks/${id}/resume`);
    return res.data;
  },

  async snoozeTask(id: number, payload: SnoozePayload): Promise<Task> {
    const res = await apiClient.post<Task>(`/tasks/${id}/snooze`, payload);
    return res.data;
  },

  async rescheduleTask(id: number, payload: ReschedulePayload): Promise<Task> {
    const res = await apiClient.post<Task>(`/tasks/${id}/reschedule`, payload);
    return res.data;
  },

  async completeTask(id: number): Promise<Task> {
    const res = await apiClient.post<Task>(`/tasks/${id}/complete`);
    return res.data;
  },

  async updateNotificationId(id: number, notificationId: string): Promise<Task> {
    const res = await apiClient.patch<Task>(`/tasks/${id}/notification`, {
      notification_id: notificationId
    });
    return res.data;
  }
};
